platform/broadcom/sonic-platform-modules-ufispace/s9300-32d/sonic_platform/fan.py

```diff
@@ -0,0 +1,181 @@
+
+    def get_speed_rpm(self):
+        """
+        Retrieves the speed of fan in RPM
+
+        Returns:
+            An integer, Speed of fan in RPM
+        """
+        rpm_speed = 0
+        if self.is_psu_fan:
+            attr = "psu_fan{}_speed_rpm".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            if self.fan_index == 1:
+                pos = "f"
+            else:
+                pos = "r"
+            attr = "fan{}_{}_speed_rpm".format(self.fantray_index, pos)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+
+        if output is None:
+            return rpm_speed
+
+        output['status'] = output['status'].rstrip()
+        if output['status'].isalpha():
+            return rpm_speed
+        else:
+            rpm_speed = int(float(output['status']))
+
+        return rpm_speed
+
+    def get_direction(self):
+        """
+        Retrieves the direction of fan
+        Returns:
+            A string, either FAN_DIRECTION_INTAKE or FAN_DIRECTION_EXHAUST
+            depending on fan direction
+        """
+        direction = self.FAN_DIRECTION_INTAKE
+        if self.is_psu_fan:
+            attr = "psu_fan{}_dir".format(self.fan_index)
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            attr = "fan{}_dir".format(self.fantray_index)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return direction
+
+        mode = output['mode']
+        val = output['status'].strip()
+        vmap = self.plugin_data['FAN']['direction'][mode]['valmap']
+   
+        if val in vmap:
+            direction = vmap[val]
+            
+        return direction
+
+    def get_presence(self):
+        """
+        Retrieves the presence of the device
+        Returns:
+            bool: True if device is present, False if not
+        """
+        presence = False
+        if self.is_psu_fan:
+            attr = "psu_present"
+            device = "PSU{}".format(self.fans_psu_index)
+        else:
+            attr = "fan{}_present".format(self.fantray_index)
+            device = "FAN-CTRL"
+
+        output = self.pddf_obj.get_attr_name_output(device, attr)
+        if not output:
+            return presence
+
+
+        mode = output['mode']
+        val = output['status'].strip()
+        vmap = self.plugin_data['FAN']['present'][mode]['valmap']
+
+        if val in vmap:
+            presence = vmap[val]
+
+        return presence
+
+    def get_status(self):
+        """
+        Retrieves the operational status of the device
+
+        Returns:
+            A boolean value, True if device is operating properly, False if not
+        """
+        speed = self.get_speed_rpm()
+        status = True if (speed != 0) else False
+        return status
+
+    def get_target_speed(self):
```

Comment:
target speed is the speed intended from the duty-cycle (or pwm) setting. I can understand of the PSU-fan intended speed is fixed but how can System fan target_speed is fixed at 65%?
Are you not supporting fan speed variations?

Reply:
Since the platform doesn't support pwm setting and the fan speed is controlled by ucd accroding to thermal sensor. Is that ok to set the target speed to crrent speed percentage to avoid warning ?